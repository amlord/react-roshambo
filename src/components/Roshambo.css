/* Roshambo */
.Roshambo {
  background: #d5d5d5;
  font-family: EncodeSansNarrowLight, helvetica, arial, sans-serif;
}

/* Roshambo Panel */
.Roshambo__container {
  max-width: 800px;
  padding: 20px;
  margin: 0 auto;
}

.Roshambo__panel {
  padding: 20px;
  background: #fff;
  border-radius: 4px;
  text-align: center;
  box-shadow: 0 3px 6px 1px rgba(0,0,0,0.14);
}

.Roshambo__title {
  margin: 0;
  line-height: 1em;
}

.Roshambo__subtitle {
  font-size: 0.8rem;
  font-weight: normal;
  margin-bottom: 20px;
  margin-top: 5px;
}

.Roshambo__game {
  font-size: 0;
}

/* Roshambo Footer */
.Roshambo__footer {
  color: #333;
  text-align: center;
  margin-top: 10px;
  font-weight: 100;
}