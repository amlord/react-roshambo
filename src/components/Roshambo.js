import React, { Component } from 'react';
import rnd  from 'pure-random';

import logo from '../logo.svg';
import './Roshambo.css';

import Player from './Player';
import Scoreboard from './Scoreboard';
import Computer from './Computer';
import { ROCK, PAPER, SCISSORS, WIN, DRAW, LOSS } from '../helpers/constants';

class Roshambo extends Component {
  constructor() {
    super();

    this.state = {
      score: {
        draw: 0,
        win: 0,
        loss: 0
      },
      history: []
    };
  }

  resetGame = () => {
    this.setState({
      score: {
        draw: 0,
        win: 0,
        loss: 0
      },
      history: []
    });
  }

  playerChoice = choice => {
    const { history } = this.state;
    const computerChoice = this.computerChoice();
    const outcome = this.gameOutcome( choice, computerChoice );

    // add the player's choice of shape to the `history` state
    history.push({
      player: choice,
      computer: computerChoice,
      outcome
    });

    // update the score state
    this.incrementScore( outcome );

    this.setState({
      history
    });
  }

  computerChoice = seed => {
    // generate random number between 1 and 3 (repeatable with `seed` for testing)
    return rnd.random( seed || rnd.genCsSeed(), 1, 3 ).value;
  }

  gameOutcome = ( playerChoice, computerChoice ) => {
    // matching choices result in a draw
    if( playerChoice === computerChoice ) {
      return DRAW;
    }

    // combinations leading to a `Player` loss
    if( ( playerChoice === ROCK && computerChoice === PAPER ) ||
        ( playerChoice === PAPER && computerChoice === SCISSORS ) ||
        ( playerChoice === SCISSORS && computerChoice === ROCK ) ) {
      return LOSS;
    }

    // anything else is a win
    return WIN;
  }

  incrementScore = outcome => {
    let { win, draw, loss } = this.state.score;

    switch( outcome ){
      case WIN:
        win += 1;
        break;
      case DRAW:
        draw += 1;
        break;
      case LOSS:
        loss += 1;
        break;
    }

    this.setState({
      score: { win, draw, loss }
    });
  }

  render() {
    return (
      <div className="Roshambo">
        <header className="Roshambo-header">
          <img src={logo} className="Roshambo-logo" alt="logo" />
          <h1 className="Roshambo-title">Welcome to Roshambo</h1>
        </header>
        <p className="Roshambo-intro">
          To get started, edit <code>src/Roshambo.js</code> and save to reload.
        </p>
        <div>
          <Player
            playerChoice={this.playerChoice}
          />
          <Scoreboard
            score={this.state.score}
            resetGame={this.resetGame}
          />
          <Computer />
        </div>
      </div>
    );
  }
}

export default Roshambo;
