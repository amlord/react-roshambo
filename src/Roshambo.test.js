import React from 'react';
import ReactDOM from 'react-dom';
import Roshambo from './Roshambo';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Roshambo />, div);
});
