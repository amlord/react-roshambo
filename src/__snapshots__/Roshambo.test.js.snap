// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Roshambo />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="Roshambo-header"
>
          <img
                    alt="logo"
                    className="Roshambo-logo"
                    src="logo.svg"
          />
          <h1
                    className="Roshambo-title"
          >
                    Welcome to Roshambo
          </h1>
</header>,
        <p
          className="Roshambo-intro"
>
          To get started, edit 
          <code>
                    src/Roshambo.js
          </code>
           and save to reload.
</p>,
      ],
      "className": "Roshambo",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <img
              alt="logo"
              className="Roshambo-logo"
              src="logo.svg"
/>,
            <h1
              className="Roshambo-title"
>
              Welcome to Roshambo
</h1>,
          ],
          "className": "Roshambo-header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "logo",
              "className": "Roshambo-logo",
              "src": "logo.svg",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Welcome to Roshambo",
              "className": "Roshambo-title",
            },
            "ref": null,
            "rendered": "Welcome to Roshambo",
            "type": "h1",
          },
        ],
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "To get started, edit ",
            <code>
              src/Roshambo.js
</code>,
            " and save to reload.",
          ],
          "className": "Roshambo-intro",
        },
        "ref": null,
        "rendered": Array [
          "To get started, edit ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "src/Roshambo.js",
            },
            "ref": null,
            "rendered": "src/Roshambo.js",
            "type": "code",
          },
          " and save to reload.",
        ],
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="Roshambo-header"
>
            <img
                        alt="logo"
                        className="Roshambo-logo"
                        src="logo.svg"
            />
            <h1
                        className="Roshambo-title"
            >
                        Welcome to Roshambo
            </h1>
</header>,
          <p
            className="Roshambo-intro"
>
            To get started, edit 
            <code>
                        src/Roshambo.js
            </code>
             and save to reload.
</p>,
        ],
        "className": "Roshambo",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                alt="logo"
                className="Roshambo-logo"
                src="logo.svg"
/>,
              <h1
                className="Roshambo-title"
>
                Welcome to Roshambo
</h1>,
            ],
            "className": "Roshambo-header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "logo",
                "className": "Roshambo-logo",
                "src": "logo.svg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Welcome to Roshambo",
                "className": "Roshambo-title",
              },
              "ref": null,
              "rendered": "Welcome to Roshambo",
              "type": "h1",
            },
          ],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "To get started, edit ",
              <code>
                src/Roshambo.js
</code>,
              " and save to reload.",
            ],
            "className": "Roshambo-intro",
          },
          "ref": null,
          "rendered": Array [
            "To get started, edit ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "src/Roshambo.js",
              },
              "ref": null,
              "rendered": "src/Roshambo.js",
              "type": "code",
            },
            " and save to reload.",
          ],
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
